src/netaddress.cpp

```diff
@@ -28,19 +28,35 @@ CNetAddr::CNetAddr()
 
 void CNetAddr::SetIP(const CNetAddr& ipIn)
 {
+    m_net = ipIn.m_net;
     memcpy(ip, ipIn.ip, sizeof(ip));
 }
 
+void CNetAddr::SetLegacyIPv6(const uint8_t ipv6[16])
```

Comment:
Any reason to pass around fixed-length data with a raw data pointer (of arbitrary length) when type safe alternatives like span or std::array (probably not applicable in this case) are available?

Reply:
The input has to be 16 bytes, in this case. Is there a way to enforce this in the type, for spans?

Reply:
Not sure, but I presumed spans could be of non-dynamic extent as well. Does this not work?

```cpp
const std::span<uint8_t, 16>& ipv6

Reply:
Yes, in this case the `[16]` is just a "signal to the programmer". In `SetRaw()` we have a bare pointer `const uint8_t*` when calling this function.

I have removed `SetRaw()` in the subsequent commit and will revisit this code.

Some experiments:

```cpp
int f1(const std::span<int, 3>& p)
{
    return p[5]; // no warning
}

int f2(const int (&p)[3])
{
    return p[5]; // warning: array index 5 is past the end of the array (which contains 3 elements) [-Warray-bounds]
}

int f3(const std::array<int, 3>& p)
{
    return p[5]; // no warning
}

int main(int, char**) {
    int a[3];
    int b[2];

    f1(a);
    f1(b); // error: no known conversion from 'int [2]' to 'const std::span<int, 3>'

    f2(a);
    f2(b); // error: no known conversion from 'int [2]' to 'const int [3]'

    f3(std::array<int, 3>{a[0], a[1], a[2]});

    return 0;
}
```

Reply:
So, in https://github.com/bitcoin/bitcoin/pull/19628 I changed the argument to `const uint8_t (&ipv6)[ADDR_IPv6_SIZE]` mostly to make the interface clean. From 3 callers 2 need an abusive typecasts because they don't have an array (one has bare pointer (the fuzz tests) and the other has `struct in6_addr`).

Feel free to comment there: https://github.com/bitcoin/bitcoin/pull/19628/files#diff-b64569708508232923e5fe3059396334R28